src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -469,66 +462,86 @@ internal void Flush()
             {
+                if (_runningFunctions.Count == 0)
+                {
+                    return;
+                }
+
+                // We only need to raise events here for functions that aren't completed.
+                // Events are raised immediately for completed functions elsewhere.
+                var runningFunctions = new List<FunctionStartedEvent>();
+                lock (_runningFunctionsSyncLock)
+                {
+                    // effectively we're pruning all the completed invocations here
+                    runningFunctions = _runningFunctions = _runningFunctions.Where(p => !p.Completed).ToList();
+                }
+
+                // we calculate concurrency here based on count, since these events are raised
+                // on a background thread, so we want the actual count for this interval, not
+                // the current count.
                 var currentTime = DateTime.UtcNow;
-                var runningFunctions = _runningFunctions.Values.ToArray();
                 foreach (var runningFunction in runningFunctions)
                 {
-                    RaiseFunctionMetricEvent(runningFunction, runningFunctions.Length, currentTime);
+                    RaiseFunctionMetricEvent(runningFunction, runningFunctions.Count, currentTime);
                 }
             }
 
-            private void RaiseFunctionMetricEvent(RunningFunctionInfo runningFunctionInfo, int concurrency, DateTime currentTime)
+            private void RaiseFunctionMetricEvent(FunctionStartedEvent runningFunctionInfo, int concurrency, DateTime currentTime)
             {
                 double executionTimespan = 0;
-                if (runningFunctionInfo.ExecutionStage == ExecutionStage.Finished)
+                ExecutionStage executionStage;
+                if (!runningFunctionInfo.Completed)
                 {
-                    executionTimespan = (runningFunctionInfo.EndTime - runningFunctionInfo.StartTime).TotalMilliseconds;
+                    executionStage = ExecutionStage.InProgress;
+                    executionTimespan = (currentTime - runningFunctionInfo.Timestamp).TotalMilliseconds;
                 }
                 else
                 {
-                    executionTimespan = (currentTime - runningFunctionInfo.StartTime).TotalMilliseconds;
+                    // regardless of the actual Failed/Succeeded status, we always raise the final event
+                    // with stage Finished
+                    executionStage = ExecutionStage.Finished;
+                    executionTimespan = runningFunctionInfo.Duration.TotalMilliseconds;
                 }
 
                 MetricsEventGenerator.LogFunctionExecutionEvent(
                     _executionId,
                     _appServiceOptions.CurrentValue.AppName,
                     concurrency,
-                    runningFunctionInfo.Name,
+                    runningFunctionInfo.FunctionMetadata.Name,
```

Comment:
I know there's a bunch of ways we write the name -- did you check this is the right one?

Reply:
I verified that we're using the same name as before. The value used before came from the same source: https://github.com/Azure/azure-functions-host/blob/826c0b520c1e81d2f3106d08711c08f1bce4aa95/src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs#L437